src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/PackageSourceMappingUtility.cs

```diff
@@ -32,73 +37,108 @@ internal static class PackageSourceMappingUtility
+                    IReadOnlyList<string> configuredSource = packageSourceMapping.GetConfiguredPackageSources(addedPackageId);
+
+                    // Top-level package was looked up.
+                    if (addedPackageId == topLevelPackageId)
                     {
-                        countCreatedTopLevelSourceMappings = 1;
-                        addedPackageIdsWithoutExistingMappings.Add(topLevelPackageId);
+                        // The top-level package is not already mapped to the selected source.
+                        if (configuredSource.Count == 0 || !configuredSource.Contains(userAction.SelectedSourceName))
+                        {
+                            countCreatedTopLevelSourceMappings++;
+                            addedPackageIdsWithoutExistingMappings.Add(topLevelPackageId);
+                        }
                     }
+                    // Transitive package was looked up.
+                    else if (configuredSource.Count == 0)
+                    {
+                        countCreatedTransitiveSourceMappings++;
+                    }
+
+                    MergePackageSourceMappings(addedSourceName, addedPackageIds, sourceMappingProvider, newAndExistingPackageSourceMappingItems);
                 }
-                // Transitive package was looked up.
-                else if (configuredSource.Count == 0)
-                {
-                    addedPackageIdsWithoutExistingMappings.Add(addedPackageId);
-                }
             }
 
-            countCreatedTransitiveSourceMappings = addedPackageIdsWithoutExistingMappings.Count - countCreatedTopLevelSourceMappings;
+            sourceMappingProvider.SavePackageSourceMappings(newAndExistingPackageSourceMappingItems);
+        }
+
+        private static string? FindSourceForPackageInGlobalPackagesFolder(
+            PackageIdentity package,
+            VersionFolderPathResolver resolver)
+        {
+            if (package is null)
+            {
+                throw new ArgumentNullException(nameof(package));
+            }
+
+            if (resolver is null)
+            {
+                throw new ArgumentNullException(nameof(resolver));
+            }
+
+            try
+            {
+                string nupkgMetadataPath = resolver.GetNupkgMetadataPath(package.Id, package.Version);
+                NupkgMetadataFile nupkgMetadata = NupkgMetadataFileFormat.Read(nupkgMetadataPath);
+
+                if (string.IsNullOrEmpty(nupkgMetadata.Source))
+                {
+                    return null;
+                }
 
-            CreateAndSavePackageSourceMappings(
-                userAction.SelectedSourceName,
-                addedPackageIdsWithoutExistingMappings,
-                sourceMappingProvider,
-                existingPackageSourceMappingSourceItems);
+                return nupkgMetadata.Source;
+            }
+            catch
```

Comment:
To prevent this code from swallowing all exceptions thrown, we should catch specific exceptions we are expecting. However, we may not need to catch any in this code. If I'm reading this correctly, the call [here](https://github.com/NuGet/NuGet.Client/blob/7252122110f1e898516f0d8471a57d92f59b7f25/src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/PackageSourceMappingUtility.cs#L109) calls into [this](https://github.com/NuGet/NuGet.Client/blob/9d6b1c6980445a84ee9fc0f114c79eb6bc32a8ba/src/NuGet.Core/NuGet.Packaging/NupkgMetadata/NupkgMetadataFileFormat.cs#L22-L25) which in turn calls [this](https://github.com/NuGet/NuGet.Client/blob/9d6b1c6980445a84ee9fc0f114c79eb6bc32a8ba/src/NuGet.Core/NuGet.Packaging/NupkgMetadata/NupkgMetadataFileFormat.cs#L29) which handles the UnauthorizedAccessException and IOException for us, returning the default. If what I'm seeing is correct, we don't need to handle those exceptions and unless we're expecting exceptions from the call to resolver.GetNupkgMetadataPath, we can remove the try/catch altogether since we're using a path that calls into a SafeRead.

Otherwise, we should catch and handle specific exceptions.

Reply:
I can understand how a method called "SafeRead" sounds like it should be "safely" handling exceptions, but it isn't actually doing that. It's more of a retry mechanism. 

I disagree that handling specific exceptions is a good idea. There are many possible types thrown here. For example, the `FileStream` constructor alone has about a dozen possible exception types it can throw. 

![image](https://github.com/NuGet/NuGet.Client/assets/49205731/c841279b-40f7-4b53-bc97-b0658a05f791)

That's in addition to bad data in the file itself: https://github.com/NuGet/NuGet.Client/blob/84cc5cc305e0c2cd528292c1230c1f80bb0d23d9/src/NuGet.Core/NuGet.Packaging/NupkgMetadata/NupkgMetadataFileFormat.cs#L49. 

How to handle the exception? In this PR, I treat it as we couldn't check the GPF. That's technically true. Whether we want a message sent to the user can be discussed (error messagebox, output window, etc). I suggest if we want this code to log something, we start with a basic fault telemetry (Filed https://github.com/NuGet/Client.Engineering/issues/2559). I don't expect this will have a high hit rate (possibly nearly zero). In fact, we didn't care about those faults before, but it is something worth reconsidering now that I've introduced another codepath for it. If we identify patterns, we can go from there developing an action plan for customers.